functions/Get-DbaSqlBuildReference.ps1

```diff
@@ -0,0 +1,210 @@
+		$LastUpdated = Get-Date -Date $IdxRef.LastUpdated
+		if ($LastUpdated -lt (Get-Date).AddDays(-45)) {
+			Write-Message -Message "Index is Stale, Last Update on: $(Get-Date -Date $LastUpdated -f s)" -Warning
+		}
+
+		function Resolve-DbaSqlBuild {
+			[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseShouldProcessForStateChangingFunctions", "")]
+			[CmdletBinding()]
+			[OutputType([System.Collections.Hashtable])]
+			Param([string]$build)
+			$LookFor = ($build.split('.')[0..2] | Foreach-Object { [convert]::ToInt32($_) }) -join '.'
+			Write-Message -Message "Looking for $LookFor" -Level 5 -Silent $Silent
+			$SqlVersion = $build.split('.')[0..1] -join '.'
+			$IdxVersion = $IdxRef.Data | Where-Object Version -like "$SqlVersion.*"
+			$Detected = @{}
+			$MatchType = 'Approximate'
+			Write-Message -Message "We have $($IdxVersion.Length) in store for this Release" -Level 5 -Silent $Silent
+			If($IdxVersion.Length -eq 0) {
+				Write-Message -Message "No info in store for this Release" -Warning -Silent $Silent
+				$Detected.Warning = "No info in store for this Release"
+			}
+			$LastVer = $IdxVersion[0]
+			foreach($el in $IdxVersion) {
+				if($null -ne $el.Name) {
+					$Detected.Name = $el.Name
+				}
+				if(Compare-DbaSqlBuildGreater -firstref $el.Version -secondref $LookFor) {
+					$MatchType = 'Approximate'
+					$Detected.Warning = "$LookFor not found, closest build we have is $($LastVer.Version)"
+					break
+				}
+				$LastVer = $el
+				if($null -ne $el.SP) {
+					$Detected.SP = $el.SP
+					$Detected.CU = $null
+				}
+				if($null -ne $el.CU) {
+					$Detected.CU = $el.CU
+				}
+				if($null -ne $el.SupportedUntil) {
+					$Detected.SupportedUntil = (Get-Date -date $el.SupportedUntil)
+				}
+				$Detected.KB = $el.KBList
+				if($el.Version -eq $LookFor) {
+					$MatchType = 'Exact'
+					break
+				}
+			}
+			return $Detected
+		}
+	}
+	PROCESS {
+		foreach ($instance in $SqlInstance) {
+			Write-Message -Message "Connecting to $instance" -Level 5 -Silent $Silent
+			try {
+				$server = Connect-SqlServer -SqlServer $instance -SqlCredential $Credential
+			} catch {
+				Stop-Function -Message "Can't connect to $instance" -Silent $Silent -Category InvalidArgument -Target $instance -Continue
+			}
+			$LookFor = $server.Version.toString()
+			$Detected = Resolve-DbaSqlBuild $LookFor
+			$object = [PSCustomObject]@{
+					SqlInstance   = $server.Name
+					InstanceName  = $server.ServiceName
+					ComputerName  = $server.NetName
+					NameLevel = $Detected.Name
+					SPLevel = $Detected.SP
+					CULevel = $Detected.CU
+					KBLevel = $Detected.KB
+					SupportedUntil = $Detected.SupportedUntil
+					MatchType = $MatchType
+					Warning = $Detected.Warning
+					Version = $LookFor
+					Instance = $server
+			}
+			Select-DefaultView -InputObject $object -Property SqlInstance, Version, NameLevel, SPLevel, CULevel, KBLevel, SupportedUntil, MatchType, Warning
+		}
+		foreach($buildstr in $Build) {
```

Comment:
Happy either way ... let's have an "internal" with the "core logic" and two main "functions", one for instances and the other for strings ? @ctrlbold  : comments ?